src/datagram.cpp

```diff
@@ -3,73 +3,257 @@
+        std::pair<uint64_t, std::vector<ngtcp2_vec>> bufs;
+        bufs.second.reserve(2);
+
+        auto& b = send_buffer.front();
+
+        prepared_datagram d{};
+        d.id = b.first;
+        d.bufs_len = 1;
 
-        for (auto& b : user_buffers)
+        oxenc::write_host_as_big(static_cast<uint16_t>(b.first), d.dgid.data());
+
+        if (_packet_splitting)
         {
-            auto& temp = bufs.emplace_back();
+            d.bufs[0].base = d.dgid.data();
+            d.bufs[0].len = 2;
+            d.bufs_len++;
+        }
+
+        d.bufs[d.bufs_len - 1].base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.second.first.data()));
+        d.bufs[d.bufs_len - 1].len = b.second.first.size();
+
+        return d;
+    }
+
+    std::optional<bstring> DatagramIO::to_buffer(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "DatagramIO handed datagram with endian swapped ID: {}", dgid);
 
-            temp.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.first.data()));
-            temp.len = b.first.size();
+        auto r = recv_buffer.receive(data, dgid);
+
+        if (r)
+            return r;
+
+        return std::nullopt;
+    }
+
+    std::optional<bstring> rotating_buffer::receive(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
+
+        {
+            std::lock_guard<std::mutex> buffer_lock(m);
+
+            auto idx = dgid >> 2;
+
+            row = (idx % bufsize) / rowsize;
+            col = idx % rowsize;
+
+            auto& b = buf.at(row).at(col);
+
+            if (not b.empty())
+            {
+#ifndef NDEBUG
+                if (enable_test_features)
+                {
+                    log::debug(log_cat, "enable_test_features is true, inducing packet loss");
+                    test_counter += 1;
+                    log::debug(log_cat, "test counter: {}", test_counter);
+                    return std::nullopt;
+                }
+                else
+                {
+                    log::debug(log_cat, "enable_test_features is false, skipping optional logic");
+                }
+#endif
+
+                log::trace(log_cat, "Pairing datagram (ID: {}) with counterpart at buffer pos [{},{}]", dgid, row, col);
+
+                bstring out{};
+                out.reserve(data.size() + b.data.size());
+
+                bool order = (b.part < 0);  // if true, we have the first part already stored
+                // must use substring or we will end up with two datagram ID's at the beginning and middle
+                out.append((order ? b.data.substr(2) : data.substr(2)));
+                out.append((order ? data.substr(2) : b.data.substr(2)));
+
+                b.clear_entry();
+
+                return out;
+            }
+
+            log::trace(log_cat, "Storing datagram (ID: {}) at buffer pos [{},{}]", dgid, row, col);
+
+            b = received_datagram{dgid, data};
```

Comment:
... that is, just before this line we can do `data.remove_prefix(2);` to discard the dgid from the data we store.

Reply:
or even earlier in connection?